Source/Dafny/Translator.cs

```diff
@@ -1266,7 +1266,7 @@ private void AddTraitParentAxioms()
         if (dd.Var.Type.IsNumericBased() || dd.Var.Type.IsBitVectorType || dd.Var.Type.IsBoolType) {
           // optimize this to only use the numeric/bitvector constraint, not the whole $Is thing on the base type
           parentConstraint = Bpl.Expr.True;
-          var udt = new UserDefinedType(dd.tok, dd.Name, (TopLevelDecl)dd, dd.TypeArgs.ConvertAll(tp => (Type)new UserDefinedType(tp)));
+          var udt = UserDefinedType.FromTopLevelDecl(dd.tok, (TopLevelDecl)dd);
```

Comment:
Just a suggestion.  Feel free to ignore.  Instead of casting `dd` to `TopLevelDecl` twice in this method, maybe you could either use parametric polymorphism to constrain the method instead.  E.g.

```
void AddRedirectingX<T>(..., T dd, ...) where T : TopLevelDecl, RedirectingTypeDecl { ... }
```

Reply:
Nice idea. I'll incorporate.